### PR TITLE
Example of DataLoaders.

### PR DESCRIPTION
**Pull request by neetkee**
### :pencil: Description
I've asked about using dataLoaders in the related issue, so here is my example of its possible implementation, using an extension of a query handler. What do you think?

Note that I also made ExecutionResult.toGraphQLResponse() function public in this PR.

### :link: Related Issues
https://github.com/ExpediaGroup/graphql-kotlin/issues/381

### CODE REVIEW

**Comment by dariuszkuc** [comment]
Thanks for providing the example! I've added default support to `SimpleQueryHandler` in https://github.com/ExpediaGroup/graphql-kotlin/pull/400.

Would you mind updating the example to use latest code?

**Comment by neetkee** [comment]
I guess we need to create DataLoaderRegistry per request, to avoid passing cached values between requests? I'm not sure how to do it with https://github.com/ExpediaGroup/graphql-kotlin/pull/400. Maybe we need to autowire some factory, that creates dataLoaderRegistry, I saw a similar way in SPQR lib: [DataLoaderRegistryFactory](https://github.com/leangen/graphql-spqr-spring-boot-starter/blob/0c2c956a0a9aea15345b210d02353c944126ae6e/graphql-spqr-spring-boot-autoconfigure/src/main/java/io/leangen/graphql/spqr/spring/autoconfigure/DataLoaderRegistryFactory.java)

**Comment by dariuszkuc** [comment]
Interesting! So data loader registry should be created per request? I've added support for this in [#410](https://github.com/ExpediaGroup/graphql-kotlin/pull/410) by creating factory that allows developers to create new data loader registry instance per query execution. Unsure how useful it is as within `graphql-kotlin` we generate schemas directly from the code based on the underlying functions and classes (where devs can implement whatever) but I think it is good to have an option to use DataLoader.

**Comment by smyrick** [comment]
👍 Plus one to this:
> Unsure how useful it is as within graphql-kotlin we generate schemas directly from the code based on the underlying functions and classes 

In graphql-java land, I can see how this is a helpful caching mechanism since you don't have a class that is run when you execute a resolver/data fetcher. But with our library you will be executing a function that can be in any scope and fetching data from a downstream service can be cached in any way you want. This feature will be something we can support through the factory but we will defer handling of everything else to the server devs.

**Comment by neetkee** [comment]
@smyrick 
> But with our library you will be executing a function that can be in any scope and fetching data from a downstream service can be cached in any way you want.

Could you give me some direction on how it can be implemented?

**Comment by smyrick** [comment]
@neetkee 


If you have this class and function as your schema object

```kotlin
class Query {
  fun getUserById(id: Int): User? {
    val result = getUserFromSource(id)
    return result
  }
}
```

the `getUserFromSource` could be any function that loads data from an API, database, or a cache system. This means that you could handle the caching layer yourself based on more inputs or configs and not solely based on the `DataLoader` implementation.

**Comment by neetkee** [comment]
If I understand correctly, we can use it to load from cache/service/etc by id, one by one, but if we want to batch our request, we need to gather all needed ids, and then execute our query.

**Comment by dariuszkuc** [comment]
> If I understand correctly, we can use it to load from cache/service/etc by id, one by one, but if we want to batch our request, we need to gather all needed ids, and then execute our query.

Yep. It all drills down to how you gather those IDs -> you need to know all of them to batch regardless whether you use data loader or not.

File: examples/spring/src/main/kotlin/com/expediagroup/graphql/examples/query/OrganizationQuery.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    @GraphQLDescription("Get all employees")
+    fun employees(): List<Employee> {
+        return employees
+    }
+}
+
+@Component("CompanyDataFetcher")
+@Scope("prototype")
+class CompanyDataFetcher : DataFetcher<CompletableFuture<Company>>, BeanFactoryAware {
```
Comment (RIP21) [inline examples/spring/src/main/kotlin/com/expediagroup/graphql/examples/query/OrganizationQuery.kt:41]: @smyrick hey. Am I understand that correctly that this thing will be called anytime somebody queries the` company` field `company`
`lateinit var Company` on Employee and ONLY the Employee, and not other classes?
Or I have to branch here or whatever?

Maybe some drawing or ASCII or draw.io graph will help me understand the connection between beans and the whole resolve cycle as I'm unsure :)

Comment (smyrick) [reply examples/spring/src/main/kotlin/com/expediagroup/graphql/examples/query/OrganizationQuery.kt:41]: @RIP21 The DataFetcher is the name of the "resolver" for graphql-java. You can see the docs here: https://www.graphql-java.com/documentation/v16/data-fetching/

This is a data fetcher for the entire `Company` class. GraphQL Kotlin normally would handle creating the different data fetchers for you, but in this case the user wanted have a `DataLoader` which is a separate GraphQL specific caching mechanism. The `DataFetcher` is invoked when ever there is a field that returns a `Company` and, using `DataLoader`s, they are cache the company info for each employee so they don't have to make multiple calls to get the the duplicate company data for all the employees

**Pull request merged**